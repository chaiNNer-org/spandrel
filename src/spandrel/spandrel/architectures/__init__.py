"""
The package containing the implementations of all supported architectures. Not necessary for most user code.
"""
