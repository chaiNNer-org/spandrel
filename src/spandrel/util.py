"""
A module containing commonly-used functionality to implement architectures.
"""

from __future__ import annotations

import functools
import inspect
import math
from typing import Any, Literal, Mapping, Protocol, TypeVar


class KeyCondition:
    """
    A condition that checks if a state dict has the given keys.
    """

    def __init__(
        self, kind: Literal["all", "any"], keys: tuple[str | KeyCondition, ...]
    ):
        self._keys = keys
        self._kind = kind

    @staticmethod
    def has_all(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("all", keys)

    @staticmethod
    def has_any(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("any", keys)

    def __call__(self, state_dict: Mapping[str, object]) -> bool:
        def _detect(key: str | KeyCondition) -> bool:
            if isinstance(key, KeyCondition):
                return key(state_dict)
            return key in state_dict

        if self._kind == "all":
            return all(_detect(key) for key in self._keys)
        else:
            return any(_detect(key) for key in self._keys)


def get_first_seq_index(state_dict: Mapping[str, object], key_pattern: str) -> int:
    """
    Returns the maximum index `i` such that `key_pattern.format(str(i))` is in `state`.

    If no such key is in state, then `-1` is returned.

    Example:
        get_first_seq_index(state, "body.{}.weight") -> -1
        get_first_seq_index(state, "body.{}.weight") -> 3
    """
    for i in range(100):
        if key_pattern.format(str(i)) in state_dict:
            return i
    return -1


def get_seq_len(state_dict: Mapping[str, object], seq_key: str) -> int:
    """
    Returns the length of a sequence in the state dict.

    The length is detected by finding the maximum index `i` such that
    `{seq_key}.{i}.{suffix}` is in `state` for some suffix.

    Example:
        get_seq_len(state, "body") -> 5
    """
    prefix = seq_key + "."

    keys: set[int] = set()
    for k in state_dict.keys():
        if k.startswith(prefix):
            index = k[len(prefix) :].split(".", maxsplit=1)[0]
            keys.add(int(index))

    if len(keys) == 0:
        return 0
    return max(keys) + 1


def get_scale_and_output_channels(x: int, input_channels: int) -> tuple[int, int]:
    """
    Returns a scale and number of output channels such that `scale**2 * out_nc = x`.

    This is commonly used for pixelshuffel layers.
    """
    # Unfortunately, we do not have enough information to determine both the scale and
    # number output channels correctly *in general*. However, we can make some
    # assumptions to make it good enough.
    #
    # What we know:
    # - x = scale * scale * output_channels
    # - output_channels is likely equal to input_channels
    # - output_channels and input_channels is likely 1, 3, or 4
    # - scale is likely 1, 2, 4, or 8

    def is_square(n: int) -> bool:
        return math.sqrt(n) == int(math.sqrt(n))

    # just try out a few candidates and see which ones fulfill the requirements
    candidates = [input_channels, 3, 4, 1]
    for c in candidates:
        if x % c == 0 and is_square(x // c):
            return int(math.sqrt(x // c)), c

    raise AssertionError(
        f"Expected output channels to be either 1, 3, or 4."
        f" Could not find a pair (scale, out_nc) such that `scale**2 * out_nc = {x}`"
    )


def store_hyperparameters(*, extra_parameters: Mapping[str, object] = {}):
    """
    Stores the hyperparameters of a class in a `hyperparameters` attribute.
    """

    def get_arg_defaults(spec: inspect.FullArgSpec) -> dict[str, Any]:
        defaults = {}
        if spec.kwonlydefaults is not None:
            defaults = spec.kwonlydefaults

        if spec.defaults is not None:
            defaults = {
                **defaults,
                **dict(zip(spec.args[-len(spec.defaults) :], spec.defaults)),
            }

        return defaults

    class WithHyperparameters(Protocol):
        hyperparameters: dict[str, Any]

    C = TypeVar("C", bound=WithHyperparameters)

    def inner(cls: type[C]) -> type[C]:
        old_init = cls.__init__

        spec = inspect.getfullargspec(old_init)
        defaults = get_arg_defaults(spec)

        if spec.varargs is not None:
            raise UserWarning(
                "Class has *args, which is not allowed in combination with @store_hyperparameters"
            )
        if spec.varkw is not None:
            raise UserWarning(
                "Class has **kwargs, which is not allowed in combination with @store_hyperparameters"
            )

        @functools.wraps(old_init)
        def new_init(self: C, **kwargs):
            # remove extra parameters from kwargs
            for k, v in extra_parameters.items():
                if k in kwargs:
                    if kwargs[k] != v:
                        raise ValueError(
                            f"Expected hyperparameter {k} to be {v}, but got {kwargs[k]}"
                        )
                    del kwargs[k]

            self.hyperparameters = {**extra_parameters, **defaults, **kwargs}
            old_init(self, **kwargs)

        cls.__init__ = new_init
        return cls

    return inner


__all__ = [
    "KeyCondition",
    "get_first_seq_index",
    "get_seq_len",
    "get_scale_and_output_channels",
    "store_hyperparameters",
]
