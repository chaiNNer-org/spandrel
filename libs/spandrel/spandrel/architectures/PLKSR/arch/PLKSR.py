from functools import partial
from typing import Literal, Sequence

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.nn.init import trunc_normal_

from spandrel.util import store_hyperparameters


# Since Pytorch's interleave is not supported by CoreML, we use this function instead for mobile conversion
def repeat_interleave(x, n):
    x = x.unsqueeze(2)
    x = x.repeat(1, 1, n, 1, 1)
    x = x.reshape(x.shape[0], x.shape[1] * n, x.shape[3], x.shape[4])
    return x


class CCM(nn.Sequential):
    "Convolutional Channel Mixer"

    def __init__(self, dim: int):
        super().__init__(
            nn.Conv2d(dim, dim * 2, 3, 1, 1),
            nn.GELU(),
            nn.Conv2d(dim * 2, dim, 1, 1, 0),
        )
        trunc_normal_(self[-1].weight, std=0.02)


class ICCM(nn.Sequential):
    "Inverted Convolutional Channel Mixer"

    def __init__(self, dim: int):
        super().__init__(
            nn.Conv2d(dim, dim * 2, 1, 1, 0),
            nn.GELU(),
            nn.Conv2d(dim * 2, dim, 3, 1, 1),
        )
        trunc_normal_(self[-1].weight, std=0.02)


class DCCM(nn.Sequential):
    "Doubled Convolutional Channel Mixer"

    def __init__(self, dim: int):
        super().__init__(
            nn.Conv2d(dim, dim * 2, 3, 1, 1),
            nn.GELU(),
            nn.Conv2d(dim * 2, dim, 3, 1, 1),
        )
        trunc_normal_(self[-1].weight, std=0.02)


class PLKConv2d(nn.Module):
    "Partial Large Kernel Convolutional Layer"

    def __init__(self, dim, kernel_size, with_idt):
        super().__init__()
        self.with_idt = with_idt
        self.conv = nn.Conv2d(dim, dim, kernel_size, 1, kernel_size // 2)
        trunc_normal_(self.conv.weight, std=0.02)
        self.idx = dim

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        if self.with_idt:
            x[:, : self.idx] = x[:, : self.idx] + self.conv(x[:, : self.idx])
        else:
            x[:, : self.idx] = self.conv(x[:, : self.idx])
        return x


class RectSparsePLKConv2d(nn.Module):
    "Rectangular Sparse Partial Large Kernel Convolutional Layer (SLaK style)"

    def __init__(self, dim, kernel_size):
        super().__init__()
        self.idx = dim
        m = kernel_size
        n = kernel_size // 3
        self.mn_conv = nn.Conv2d(dim, dim, (m, n), 1, (m // 2, n // 2))
        self.nm_conv = nn.Conv2d(dim, dim, (n, m), 1, (n // 2, m // 2))
        self.nn_conv = nn.Conv2d(dim, dim, (n, n), 1, (n // 2, n // 2))

        trunc_normal_(self.mn_conv.weight, std=0.02)
        trunc_normal_(self.nm_conv.weight, std=0.02)
        trunc_normal_(self.nn_conv.weight, std=0.02)

    def forward(
        self, x: torch.Tensor
    ) -> torch.Tensor:  # No reparametrization since this is for a ablative study
        if self.training:
            x1, x2 = x[:, : self.idx], x[:, self.idx :]
            x1 = self.mn_conv(x1) + self.nm_conv(x1) + self.nn_conv(x1)
            return torch.cat([x1, x2], dim=1)

        else:
            x[:, : self.idx] = (
                self.mn_conv(x[:, : self.idx])
                + self.nm_conv(x[:, : self.idx])
                + self.nn_conv(x[:, : self.idx])
            )
            return x


class SparsePLKConv2d(nn.Module):
    "Sparse Partial Large Kernel Convolutional Layer (RepLKNet and UniRepLKNet style)"

    def __init__(
        self,
        dim,
        max_kernel_size,
        sub_kernel_sizes,
        dilations,
        use_max_kernel,
        with_idt,
    ):
        super().__init__()
        self.use_max_kernel = use_max_kernel
        self.max_kernel_size = max_kernel_size
        for k, d in zip(sub_kernel_sizes, dilations):
            m_k = self._calc_rep_kernel_size(k, d)
            if m_k > self.max_kernel_size:
                self.max_kernel_size = m_k
        self.with_idt = with_idt

        convs = [
            nn.Conv2d(
                dim, dim, sub_kernel_size, 1, (sub_kernel_size // 2) * d, dilation=d
            )
            for sub_kernel_size, d in zip(sub_kernel_sizes, dilations)
        ]
        if use_max_kernel:
            convs.append(
                nn.Conv2d(dim, dim, self.max_kernel_size, 1, self.max_kernel_size // 2)
            )
        self.convs = nn.ModuleList(convs)
        for m in self.convs:
            trunc_normal_(m.weight, std=0.02)
        self.idx = dim
        self.is_convert = False

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        if self.is_convert:
            x[:, : self.idx, :, :] = self.conv(x[:, : self.idx, :, :])
            return x
        else:
            x1, x2 = torch.split(x, [self.idx, x.size(1) - self.idx], dim=1)
            if self.with_idt:
                out = x1
            else:
                out = 0.0
            for conv in self.convs:
                out = out + conv(x1)
            return torch.cat([out, x2], dim=1)  # type: ignore

    @staticmethod
    def _calc_rep_kernel_size(ks, dilation):
        return (ks - 1) * dilation + 1

    @staticmethod
    def _get_origin_kernel(kernel, dilation=1, p=0):
        I = torch.ones((1, 1, 1, 1)).to(kernel.device)  # noqa: E741
        if kernel.size(1) == 1:  # Depth-wise Convolution
            dilated = F.conv_transpose2d(kernel, I, stride=dilation)
        else:
            slices = []  # Dense or Group
            for i in range(kernel.size(1)):
                dilated = F.conv_transpose2d(
                    kernel[:, i : i + 1, :, :], I, stride=dilation
                )
                slices.append(dilated)
            dilated = torch.cat(slices, dim=1)

        # Pad boundary
        if p != 0:
            dilated = F.pad(dilated, (p, p, p, p))
        return dilated

    @staticmethod
    def _dwc_to_dense(kernel):
        n_groups = kernel.size(0)

        kernels = []
        for g in range(n_groups):
            kernels.append(
                torch.cat(
                    [
                        kernel[g : (g + 1)]
                        if g == i
                        else torch.zeros_like(kernel[g : (g + 1)])
                        for i in range(n_groups)
                    ],
                    dim=1,
                )
            )
        return torch.cat(kernels, dim=0)


class EA(nn.Module):
    "Element-wise Attention"

    def __init__(self, dim: int):
        super().__init__()
        self.f = nn.Sequential(nn.Conv2d(dim, dim, 3, 1, 1), nn.Sigmoid())
        trunc_normal_(self.f[0].weight, std=0.02)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x * self.f(x)


class PLKBlock(nn.Module):
    def __init__(
        self,
        dim: int,
        # CCM Rep options
        ccm_type: Literal["CCM", "ICCM", "DCCM"],
        # LK Options
        max_kernel_size: int,
        split_ratio: float,
        lk_type: Literal["PLK", "SparsePLK", "RectSparsePLK"] = "PLK",
        # Sparse Rep options
        use_max_kernel: bool = False,
        sparse_kernels: Sequence[int] = [5, 5, 5],
        sparse_dilations: Sequence[int] = [2, 3, 4],
        with_idt: bool = False,
        # EA ablation
        use_ea: bool = True,
    ):
        super().__init__()

        # Local Texture
        if ccm_type == "CCM":
            self.channe_mixer = CCM(dim)
        elif ccm_type == "ICCM":
            self.channe_mixer = ICCM(dim)
        elif ccm_type == "DCCM":
            self.channe_mixer = DCCM(dim)
        else:
            raise ValueError(f"Unknown CCM type: {ccm_type}")

        # Long-range Dependency
        pdim = int(dim * split_ratio)
        if lk_type == "PLK":
            self.lk = PLKConv2d(pdim, max_kernel_size, with_idt)
        elif lk_type == "SparsePLK":
            self.lk = SparsePLKConv2d(
                pdim,
                max_kernel_size,
                sparse_kernels,
                sparse_dilations,
                use_max_kernel,
                with_idt,
            )
        elif lk_type == "RectSparsePLK":
            self.lk = RectSparsePLKConv2d(pdim, max_kernel_size)
        else:
            raise ValueError(f"Unknown LK type: {lk_type}")

        # Instance-dependent modulation
        if use_ea:
            self.attn = EA(dim)
        else:
            self.attn = nn.Identity()

        # Refinement
        self.refine = nn.Conv2d(dim, dim, 1, 1, 0)
        trunc_normal_(self.refine.weight, std=0.02)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x_skip = x
        x = self.channe_mixer(x)
        x = self.lk(x)
        x = self.attn(x)
        x = self.refine(x)
        return x + x_skip


@store_hyperparameters()
class PLKSR(nn.Module):
    hyperparameters = {}

    def __init__(
        self,
        dim: int = 64,
        n_blocks: int = 28,
        upscaling_factor: int = 4,
        # CCM options
        ccm_type: Literal["CCM", "ICCM", "DCCM"] = "CCM",
        # LK Options
        kernel_size: int = 17,
        split_ratio: float = 0.25,
        lk_type: Literal["PLK", "SparsePLK", "RectSparsePLK"] = "PLK",
        # LK Rep options
        use_max_kernel: bool = False,
        sparse_kernels: Sequence[int] = [5, 5, 5, 5],
        sparse_dilations: Sequence[int] = [1, 2, 3, 4],
        with_idt: bool = False,
        # EA ablation
        use_ea: bool = True,
    ):
        super().__init__()

        self.feats = nn.Sequential(
            *[nn.Conv2d(3, dim, 3, 1, 1)]
            + [
                PLKBlock(
                    dim,
                    ccm_type,
                    kernel_size,
                    split_ratio,
                    lk_type,
                    use_max_kernel,
                    sparse_kernels,
                    sparse_dilations,
                    with_idt,
                    use_ea,
                )
                for _ in range(n_blocks)
            ]
            + [nn.Conv2d(dim, 3 * upscaling_factor**2, 3, 1, 1)]
        )
        trunc_normal_(self.feats[0].weight, std=0.02)
        trunc_normal_(self.feats[-1].weight, std=0.02)

        self.to_img = nn.PixelShuffle(upscaling_factor)

        self.repeat_op = partial(
            torch.repeat_interleave, repeats=upscaling_factor**2, dim=1
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.feats(x) + self.repeat_op(x)
        x = self.to_img(x)
        return x
