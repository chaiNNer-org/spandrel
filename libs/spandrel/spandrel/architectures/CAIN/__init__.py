from __future__ import annotations

import math
from typing import Union

from typing_extensions import override

from spandrel.util import KeyCondition

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.cain import CAIN
from .arch.cain_encdec import CAIN_EncDec
from .arch.cain_noca import CAIN_NoCA

_CainModel = Union[CAIN, CAIN_NoCA, CAIN_EncDec]


class CAINArch(Architecture[_CainModel]):
    def __init__(self) -> None:
        super().__init__(
            id="CAIN",
            detect=KeyCondition.has_any(
                # CAIN
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_NoCA
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_EncDec
                KeyCondition.has_all(
                    "encoder.body.0.conv.weight",
                    "encoder.body.6.conv.weight",
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                    "decoder.body.1.body.0.conv.weight",
                    "decoder.body.1.body.2.conv.bias",
                    "decoder.body.3.body.0.conv.weight",
                    "decoder.body.3.body.2.conv.bias",
                    "decoder.body.5.body.0.conv.weight",
                    "decoder.body.5.body.2.conv.bias",
                ),
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[_CainModel]:
        tags: list[str] = []

        if "encoder.body.0.conv.weight" in state_dict:
            start_filters = 32
            up_mode = "shuffle"

            start_filters = state_dict["decoder.body.5.body.0.conv.weight"].shape[0]

            if "decoder.body.0.upconv.0.conv.weight" in state_dict:
                up_mode = "shuffle"
            elif "decoder.body.0.upconv.weight" in state_dict:
                up_mode = "transpose"
            else:
                up_mode = "direct"

            model = CAIN_EncDec(
                start_filters=start_filters,
                up_mode=up_mode,
            )
            tags.extend(("EncDec", f"{start_filters}sf"))
        else:
            depth = 3

            # detect
            # n_feats = 3 * (4**depth)
            n_feats = state_dict["encoder.interpolate.headConv.weight"].shape[1]
            depth = math.isqrt(math.isqrt(n_feats // 3))

            if (
                "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight"
                in state_dict
            ):
                model = CAIN(
                    depth=depth,
                )
            else:
                model = CAIN_NoCA(
                    depth=depth,
                )
                tags.append("NoCA")

            tags.append(f"{depth}depth")

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration",
            tags=tags,
            supports_half=True,
            supports_bfloat16=True,
            scale=1,
            input_channels=3,
            output_channels=3,
        )
