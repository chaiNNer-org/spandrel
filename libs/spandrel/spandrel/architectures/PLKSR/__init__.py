from __future__ import annotations

import math
from typing import Union

from typing_extensions import override

from spandrel.util import KeyCondition, get_seq_len

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.PLKSR import PLKSR
from .arch.RealPLKSR import RealPLKSR

_PLKSR = Union[PLKSR, RealPLKSR]


class PLKSRArch(Architecture[_PLKSR]):
    def __init__(self) -> None:
        super().__init__(
            id="PLKSR",
            detect=KeyCondition.has_all(
                "feats.0.weight",
                "feats.1.lk.conv.weight",
                "feats.1.refine.weight",
                KeyCondition.has_any(
                    "feats.1.channe_mixer.0.weight",
                    "feats.1.channel_mixer.0.weight",
                ),
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[_PLKSR]:
        dim = 64
        n_blocks = 28
        scale = 4
        kernel_size = 17
        split_ratio = 0.25
        use_ea = True

        # RealPLKSR only
        norm_groups = 4  # un-detectable
        dropout = 0  # un-detectable

        dim = state_dict["feats.0.weight"].shape[0]

        total_feat_layers = get_seq_len(state_dict, "feats")
        scale = math.isqrt(
            state_dict[f"feats.{total_feat_layers - 1}.weight"].shape[0] // 3
        )

        kernel_size = state_dict["feats.1.lk.conv.weight"].shape[2]
        split_ratio = state_dict["feats.1.lk.conv.weight"].shape[0] / dim

        use_ea = "feats.1.attn.f.0.weight" in state_dict

        # Yes, the normal version has this typo.
        if "feats.1.channe_mixer.0.weight" in state_dict:
            n_blocks = total_feat_layers - 2

            mixer_0_shape = state_dict["feats.1.channe_mixer.0.weight"].shape[2]
            mixer_2_shape = state_dict["feats.1.channe_mixer.2.weight"].shape[2]
            if mixer_0_shape == 3 and mixer_2_shape == 1:
                ccm_type = "CCM"
            elif mixer_0_shape == 3 and mixer_2_shape == 3:
                ccm_type = "DCCM"
            elif mixer_0_shape == 1 and mixer_2_shape == 3:
                ccm_type = "ICCM"
            else:
                raise ValueError("Unknown CCM type")
            more_tags = [ccm_type]

            model = PLKSR(
                dim=dim,
                upscaling_factor=scale,
                n_blocks=n_blocks,
                kernel_size=kernel_size,
                split_ratio=split_ratio,
                use_ea=use_ea,
                ccm_type=ccm_type,
            )
        # and RealPLKSR doesn't. This makes it really convenient to detect.
        elif "feats.1.channel_mixer.0.weight" in state_dict:
            more_tags = ["Real"]

            n_blocks = total_feat_layers - 3
            model = RealPLKSR(
                dim=dim,
                upscaling_factor=scale,
                n_blocks=n_blocks,
                kernel_size=kernel_size,
                split_ratio=split_ratio,
                use_ea=use_ea,
                norm_groups=norm_groups,
                dropout=dropout,
            )
        else:
            raise ValueError("Unknown model type")

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration" if scale == 1 else "SR",
            tags=[f"{dim}dim", f"{n_blocks}nb", f"{kernel_size}ks", *more_tags],
            supports_half=False,
            supports_bfloat16=True,
            scale=scale,
            input_channels=3,
            output_channels=3,
        )
