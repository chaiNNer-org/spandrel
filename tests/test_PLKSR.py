from spandrel.architectures.PLKSR import PLKSR, PLKSRArch, RealPLKSR

from .util import (
    ModelFile,
    TestImage,
    assert_image_inference,
    assert_loads_correctly,
    assert_size_requirements,
    disallowed_props,
    skip_if_unchanged,
)

skip_if_unchanged(__file__)


def test_load():
    assert_loads_correctly(
        PLKSRArch(),
        # PLKSR
        lambda: PLKSR(),
        lambda: PLKSR(dim=32),
        lambda: PLKSR(dim=96),
        lambda: PLKSR(n_blocks=6),
        lambda: PLKSR(n_blocks=35),
        lambda: PLKSR(upscaling_factor=2),
        lambda: PLKSR(upscaling_factor=6),
        lambda: PLKSR(ccm_type="DCCM"),
        lambda: PLKSR(ccm_type="CCM"),
        lambda: PLKSR(ccm_type="ICCM"),
        lambda: PLKSR(lk_type="PLK", kernel_size=9),
        lambda: PLKSR(lk_type="PLK", kernel_size=27),
        lambda: PLKSR(lk_type="PLK", split_ratio=0.5),
        lambda: PLKSR(lk_type="PLK", split_ratio=0.75),
        lambda: PLKSR(lk_type="RectSparsePLK", kernel_size=9),
        lambda: PLKSR(lk_type="RectSparsePLK", kernel_size=27),
        lambda: PLKSR(lk_type="RectSparsePLK", split_ratio=0.5),
        lambda: PLKSR(lk_type="RectSparsePLK", split_ratio=0.75),
        lambda: PLKSR(lk_type="SparsePLK", split_ratio=0.5),
        lambda: PLKSR(lk_type="SparsePLK", split_ratio=0.75),
        lambda: PLKSR(use_ea=False),
        # RealPLKSR
        lambda: RealPLKSR(),
        lambda: RealPLKSR(dim=32),
        lambda: RealPLKSR(dim=96),
        lambda: RealPLKSR(n_blocks=6),
        lambda: RealPLKSR(n_blocks=35),
        lambda: RealPLKSR(upscaling_factor=2),
        lambda: RealPLKSR(upscaling_factor=6),
        lambda: RealPLKSR(kernel_size=9),
        lambda: RealPLKSR(kernel_size=27),
        lambda: RealPLKSR(split_ratio=0.5),
        lambda: RealPLKSR(split_ratio=0.75),
        lambda: RealPLKSR(use_ea=False),
    )


def test_size_requirements():
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/12ek1vitEporWc5qqaYo6AMy0-RYlRqu8/view",
        name="4x_realplksr_mssim_pretrain.pth",
    )
    assert_size_requirements(file.load_model())

    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1PA3QElJYlgpPYKl0zQ9_D1pnuYfC1vtt/view",
        name="PLKSR_X2_DIV2K.pth",
    )
    assert_size_requirements(file.load_model())


def test_PLKSR_official_x4(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1YtoOjK7vsfVrHFqWGvdjFNcOay8WUuJ7/view",
        name="PLKSR_X4_DIV2K.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, PLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )


def test_PLKSR_official_tiny_x4(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1d8d_6TF0SrEMiX1jrnLqKnPdKjDJahOK/view",
        name="PLKSR_tiny_X4_DIV2K.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, PLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )


def test_PLKSR_official_x3(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1W8phbKFTOYL-AnlMJnjx2NWDHZY8jVWW/view",
        name="PLKSR_X3_DIV2K.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, PLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )


def test_PLKSR_official_x2(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1PA3QElJYlgpPYKl0zQ9_D1pnuYfC1vtt/view",
        name="PLKSR_X2_DIV2K.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, PLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )


def test_RealPLKSR_4x(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/12ek1vitEporWc5qqaYo6AMy0-RYlRqu8/view",
        name="4x_realplksr_mssim_pretrain.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, RealPLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )


def test_RealPLKSR_2x(snapshot):
    file = ModelFile.from_url(
        "https://drive.google.com/file/d/1GAdf5VOqYa5ntswT9sYsKKZ2Z7OQp7gO/view",
        name="2x_realplksr_mssim_pretrain.pth",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, RealPLKSR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )
