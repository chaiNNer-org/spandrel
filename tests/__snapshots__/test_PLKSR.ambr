# serializer version: 1
# name: test_PLKSR_official_tiny_x4
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=4,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '12nb',
      '13ks',
      'DCCM',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
# name: test_PLKSR_official_x2
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=2,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '28nb',
      '17ks',
      'DCCM',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
# name: test_PLKSR_official_x3
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=3,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '28nb',
      '17ks',
      'DCCM',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
# name: test_PLKSR_official_x4
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=4,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '28nb',
      '17ks',
      'DCCM',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
# name: test_RealPLKSR_2x
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=2,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '28nb',
      '17ks',
      'Real',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
# name: test_RealPLKSR_4x
  ImageModelDescriptor(
    architecture=PLKSRArch(
      id='PLKSR',
      name='PLKSR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=4,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      '64dim',
      '28nb',
      '17ks',
      'Real',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
